File: 19th/코딩테스트 심화반/week4/19238번_스타트 택시.cpp
#include <iostream>
#include <queue>

using namespace std;

int map[21][21];

// 플러드필
int dr[4] = { 0, 1, 0, -1 };
int dc[4] = { 1, 0, -1, 0 };

// 각 승객의 목적지
pair<int, int> end_pos[401];

int main() {
    int N, M, fuel;
    int r, c;
    int s_r, s_c, e_r, e_c;

    cin >> N >> M >> fuel;

    int input;
    for (int i = 1; i <= N; i++) {
        for (int j = 1; j <= N; j++) {
            cin >> input;
            if (input == 1) map[i][j] = -1; // 벽을 -1로 표현
            else map[i][j] = 0;
        }
    }

    cin >> r >> c;

    for (int i = 1; i <= M; i++) {
        cin >> s_r >> s_c >> e_r >> e_c;
        map[s_r][s_c] = i; // n번 승객의 출발지를 n으로 표현 
        end_pos[i] = { e_r, e_c }; // 각 승객의 목적지가 겹칠 수 있으므로, 따로 저장
    }

    bool find_path = false;
    while (M--) {
        // 현재 위치 -> n번 승객 출발지
        // n번 승객 출발지 -> n번 승객 목적지
        // 두 번의 플러드필(BFS)
        bool visited1[21][21] = {};
        bool visited2[21][21] = {};
        queue<pair<pair<int, int>, int>> q1; // 위치(r, c)와 이동 거리 저장
        queue<pair<pair<int, int>, int>> q2; // 위치(r, c)와 이동 거리 저장
        int customer_num; // 현재 탑승중인 승객 번호

        // 첫 번째 플러드필 (현재 위치 -> n번 승객 출발지)
        find_path = false; // 승객을 찾았는지 판단하기 위함  
        int used_fuel1 = 401; // 승객을 찾는데 사용한 연료
        
        if (1 <= map[r][c]) { // 만약 현재 위치가 n번 승객의 출발지일 경우
            find_path = true;
            customer_num = map[r][c];

            used_fuel1 = 0;
        }
        else { // 빈칸이라면 플러드필 시작
            q1.push({ {r, c}, 0 });
            visited1[r][c] = true;
        }

        while (!q1.empty()) {
            auto pos = q1.front().first;
            int depth = q1.front().second;
            q1.pop();

            // 만약 연료를 다 썼거나, 다른 승객보다 멀리 위치할 경우 break
            if (fuel <= depth || depth >= used_fuel1) break; 

            // 사방으로 이동 시작
            for (int i = 0; i < 4; i++) { 
                int nr = pos.first + dr[i];
                int nc = pos.second + dc[i];
                if (1 <= nr && nr <= N && 1 <= nc && nc <= N) {
                    if (map[nr][nc] == -1) continue; // 벽일 경우

                    if (!visited1[nr][nc]) {
                        if (1 <= map[nr][nc]) { // n번 승객의 출발지일 경우
                            if (!find_path) { // 찾은 승객이 없을 경우
                                find_path = true;
                                customer_num = map[nr][nc];
                                r = nr;
                                c = nc;

                                used_fuel1 = depth + 1;
                            }
                            else { // 찾은 승객이 있을 경우
                                if (nr < r) { // 행 번호가 작은순
                                    customer_num = map[nr][nc];
                                    r = nr;
                                    c = nc;
                                }
                                else if (nr == r) {
                                    if (nc < c) { // 열 번호가 작은순
                                        customer_num = map[nr][nc];
                                        r = nr;
                                        c = nc;
                                    }
                                }
                            }
                        }
                        else { // 빈칸일 경우
                            q1.push({ {nr, nc}, depth + 1 });
                            visited1[nr][nc] = true;
                        }
                        
                    }
                }
            }
        }

        if (!find_path) break;
        map[r][c] = 0; // 해당 위치를 빈칸으로 만듦
        fuel -= used_fuel1;

        // 두 번째 플러드필 (n번 승객 출발지 -> n번 승객 목적지)
        find_path = false; // 승객을 도착지까지 이동시켰는지 판단하기 위함  
        int used_fuel2 = 401; // 승객을 도착지에 이동시키는데 사용한 연료

        // 플러드필 시작
        q2.push({ {r, c}, 0 });
        visited2[r][c] = true;

        while (!q2.empty()) {
            auto pos = q2.front().first;
            int depth = q2.front().second;
            q2.pop();

            // 만약 연료를 다 썼거나, 이미 경로를 찾았을 경우 break
            if (fuel <= depth || find_path) break;

            // 사방으로 이동 시작
            for (int i = 0; i < 4; i++) {
                int nr = pos.first + dr[i];
                int nc = pos.second + dc[i];
                if (1 <= nr && nr <= N && 1 <= nc && nc <= N) {
                    if (map[nr][nc] == -1) continue; // 벽일 경우

                    if (!visited2[nr][nc]) { 
                        if (nr == end_pos[customer_num].first && nc == end_pos[customer_num].second) { // 승객의 도착지일 경우
                            find_path = true;
                            r = nr;
                            c = nc;

                            used_fuel2 = depth + 1;
                        }
                        else { // 빈칸일 경우
                            q2.push({ {nr, nc}, depth + 1 });
                            visited2[nr][nc] = true;
                        } 
                    }
                }
            }
        }

        if (!find_path) break;
        fuel -= used_fuel2;

        fuel += used_fuel2 * 2;
    }

    if (!find_path) cout << -1;
    else cout << fuel;
}

File: 19th/코딩테스트 심화반/week4/19238번_스타트 택시.py
from collections import deque

# 입력: n: 지도 크기, m: 승객 수, fuel: 초기 연료
n, m, fuel = map(int, input().split())
b_map = [[0] * (n + 1)]  # 1-indexed를 위한 더미 행 추가

# 지도 정보 입력 (1은 벽이므로 -1로 치환)
for _ in range(n):
    row = list(map(int, input().split()))
    b_map.append([0] + [(-1 if x == 1 else x) for x in row])

# 택시 시작 위치
r, c = map(int, input().split())

# 승객 정보 저장 및 시작 위치에 번호 표시
passenger = []
for i in range(m):
    s_r, s_c, e_r, e_c = map(int, input().split())
    passenger.append((e_r, e_c))          # 목적지만 저장
    b_map[s_r][s_c] = i + 1               # 출발 위치에 승객 번호 등록

# 방향 벡터 (우, 하, 좌, 상)
dr = [0, 1, 0, -1]
dc = [1, 0, -1, 0]

find_path = False

# 승객 수만큼 반복 (각 승객을 찾아서 목적지까지 이동)
for _ in range(m):
    q1 = deque()
    q2 = deque()
    visited1 = [[0 for _ in range(21)] for _ in range(21)]
    visited2 = [[0 for _ in range(21)] for _ in range(21)]

    find_path = False
    used_fuel1 = 401  # 승객 찾을 때 연료 사용량 초기화

    # 현재 위치에 승객이 있으면 바로 태움
    if b_map[r][c] >= 1:
        find_path = True
        customer_num = b_map[r][c]
        used_fuel1 = 0
    else:
        # BFS 시작
        q1.append(((r, c), 0))
        visited1[r][c] = 1

    # 승객 찾기 BFS
    while q1:
        pos, depth = q1.popleft()

        if (fuel <= depth or depth >= used_fuel1):
            break  # 연료 부족하거나 최단거리보다 멀면 중단

        for i in range(4):
            nr, nc = pos[0] + dr[i], pos[1] + dc[i]

            if (1 <= nr <= n and 1 <= nc <= n and b_map[nr][nc] != -1 and not visited1[nr][nc]):
                if 1 <= b_map[nr][nc]:
                    if not find_path:
                        # 처음 찾은 승객이면 등록
                        find_path = True
                        customer_num = b_map[nr][nc]
                        r = nr
                        c = nc
                        used_fuel1 = depth + 1
                    else:
                        # 거리 같을 경우 행, 열 우선순위 비교
                        if nr < r or (nr == r and nc < c):
                            customer_num = b_map[nr][nc]
                            r = nr
                            c = nc
                else:
                    # 빈 칸이면 계속 탐색
                    q1.append(((nr, nc), depth + 1))
                    visited1[nr][nc] = True

    if not find_path:
        break  # 태울 승객 못 찾은 경우 실패
    b_map[r][c] = 0      # 승객 태웠으므로 지도에서 제거
    fuel -= used_fuel1   # 연료 차감

    find_path = False
    used_fuel2 = 401     # 도착지까지 연료 사용량 초기화

    # 승객 목적지까지 BFS
    q2.append(((r, c), 0))
    visited2[r][c] = True

    while q2:
        pos, depth = q2.popleft()

        if (fuel <= depth or find_path):
            break  # 연료 부족하거나 이미 도착한 경우 중단

        for i in range(4):
            nr, nc = pos[0] + dr[i], pos[1] + dc[i]

            if (1 <= nr <= n and 1 <= nc <= n and b_map[nr][nc] != -1 and not visited2[nr][nc]):
                destination_r, destination_c = passenger[customer_num - 1]

                if nr == destination_r and nc == destination_c:
                    # 목적지 도착
                    find_path = True
                    r = nr
                    c = nc
                    used_fuel2 = depth + 1
                else:
                    q2.append(((nr, nc), depth + 1))
                    visited2[nr][nc] = True

    if not find_path:
        break  # 목적지 못 감
    fuel -= used_fuel2           # 연료 차감
    fuel += used_fuel2 * 2       # 연료 보충

# 실패 시 -1 출력
if not find_path:
    print(-1)
else:
    print(fuel)
